rfc/Errors-Extension-Specification.md

```diff
@@ -0,0 +1,123 @@
+# Subgraph Errors Extension Specification
+
+## Problem
+
+For security reasons, we're currently not exposing Subgraph Errors to the client.
+Subgraphs might leak sensitive information in the error object, e.g. through the `extensions` field.
+
+What we've learned from our users is that they would like to leverage the `extensions` field to provide more context to the client,
+e.g. by adding a `code` field to the error object.
+
+A response object might look like this:
+
+```json
+{
+  "errors": [
+    {
+      "message": "Cannot query field 'foo' on type 'Bar'",
+      "path": [
+        "foo"
+      ],
+      "extensions": {
+        "code": "GRAPHQL_VALIDATION_FAILED"
+      }
+    }
+  ]
+}
+```
+
+With this information, the client can now handle the error more gracefully.
+
+However, this leads to the problem that Subgraph owners might send "anything" in the `extensions` field,
+which might leak sensitive information.
+
+Another problem is that users would like to standardize the error codes across different Subgraphs,
+e.g. `GRAPHQL_VALIDATION_FAILED` should always mean the same thing,
+and there should be agreement across all Subgraph owners on what codes to use.
+
+Additional fields in the `extensions` object could also be allowed, e.g. `message` or `severity`,
+but it's important that all Supergraph contributors agree on what fields and values are allowed.
+
+## Solution
+
+We propose to introduce the ability to define a schema for the `extensions` field in the Supergraph.
+As the response encoding is JSON, we can leverage JSON Schema to define the schema for the error object.
+
+The schema can be attached at the Supergraph level using the `wgc` cli tool using the following command:
+
+```bash
+npx wgc federated-graph create production -r http://router.example.com/graphql --label-matcher team=A department=backend --errors-extensions-schema ./errors-extensions-schema.json
+```
+
+Example of updating the schema for an existing Supergraph:
+
+```bash
+npx wgc federated-graph update production --errors-extensions-schema ./errors-extensions-schema.json
+```
+
+The schema file should look like this:
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "properties": {
+    "code": {
+      "type": "string",
+      "enum": [
+        "GRAPHQL_VALIDATION_FAILED",
+        "GRAPHQL_EXECUTION_FAILED"
+      ]
+    }
+  },
+  "required": [
+    "code"
+  ],
+  "additionalItems": false
+}
+```
+
+The `additionalItems` field is set to `false` to ensure that only the fields defined in the schema are allowed.
+
+In addition to just defining the schema,
+we also propose to add an additional option to define the behavior when the schema validation fails.
+
+When the JSON Schema validation fails for an `extensions` field,
```

Comment:
when is this schema validation happening?

Reply:
At runtime when an error is rendered by the Router

Reply:
I'm a bit worried about how we enforce the correct emission of errors from the subgraphs. Ideally at compile time we want to ensure that subgraphs are emitting errors in the correct format.

I know this is a language and subgraph server specific issue though, so maybe the solution to this can't necessarily live here.

Reply:
Here's how I think this would be possible.
The JSON Schema is stored in the Schema Registry.
At build time, the Subgraph downloads the Schema using a `wgc` command.
The Subgraph codebase can then generate types/models from the Subgraph Error Extensions Schema.
A mismatch will result in a compiler error in most frameworks, given that the Subgraph uses the errors.
This would still require the developers to opt-in into the tooling, but if they do, we can catch errors at build time.
WDYT?